File: python/examples/porta_example.py
"""
Porta Cipher Example

This example demonstrates the Porta cipher, which is a self-reciprocal polyalphabetic
substitution cipher that uses alphabet pairs. Each letter in the keyword determines
which alphabet pair to use for encryption.
"""

from cryptology.classical.substitution.polyalphabetic.porta import (
    encrypt, decrypt, generate_random_key, generate_key_for_text, encrypt_with_random_key
)


def main():
    print("=" * 60)
    print("PORTA CIPHER EXAMPLE")
    print("=" * 60)
    
    # Example 1: Basic encryption/decryption
    print("\n1. Basic Encryption/Decryption")
    print("-" * 40)
    
    plaintext = "HELLO WORLD"
    key = "KEY"
    
    print(f"Plaintext: {plaintext}")
    print(f"Key:       {key}")
    
    encrypted = encrypt(plaintext, key)
    decrypted = decrypt(encrypted, key)
    
    print(f"Encrypted: {encrypted}")
    print(f"Decrypted: {decrypted}")
    print(f"Success:   {plaintext == decrypted}")
    
    # Example 2: Self-reciprocal property demonstration
    print("\n2. Self-Reciprocal Property")
    print("-" * 40)
    
    plaintext = "SECRET MESSAGE"
    key = "PORTACIPHER"
    
    print(f"Plaintext: {plaintext}")
    print(f"Key:       {key}")
    
    encrypted = encrypt(plaintext, key)
    print(f"Encrypted: {encrypted}")
    
    # Demonstrate self-reciprocal property
    encrypted_again = encrypt(encrypted, key)
    print(f"Encrypt encrypted text: {encrypted_again}")
    print(f"Self-reciprocal: {encrypted_again == plaintext}")
    
    # Example 3: Alphabet pairs demonstration
    print("\n3. Alphabet Pairs Demonstration")
    print("-" * 40)
    
    print("Porta cipher uses 13 alphabet pairs:")
    pairs = [
        ("A", "N"), ("B", "O"), ("C", "P"), ("D", "Q"), ("E", "R"),
        ("F", "S"), ("G", "T"), ("H", "U"), ("I", "V"), ("J", "W"),
        ("K", "X"), ("L", "Y"), ("M", "Z")
    ]
    
    for i, (a, b) in enumerate(pairs):
        print(f"Pair {i+1:2d}: {a} ↔ {b}")
    
    print("\nTesting individual pairs:")
    test_pairs = [("A", "N"), ("C", "P"), ("E", "R"), ("G", "T")]
    
    for a, b in test_pairs:
        # Test with key 'A' (uses first pair)
        encrypted_a = encrypt(a, "A")
        encrypted_b = encrypt(b, "A")
        print(f"Key 'A': {a} -> {encrypted_a}, {b} -> {encrypted_b}")
        
        # Test with key 'B' (same pair, different mapping)
        encrypted_a_b = encrypt(a, "B")
        encrypted_b_b = encrypt(b, "B")
        print(f"Key 'B': {a} -> {encrypted_a_b}, {b} -> {encrypted_b_b}")
    
    # Example 4: Key repetition for longer messages
    print("\n4. Key Repetition")
    print("-" * 40)
    
    long_text = "THIS IS A LONG MESSAGE THAT REQUIRES KEY REPETITION"
    short_key = "AB"
    
    print(f"Text:      {long_text}")
    print(f"Key:       {short_key}")
    
    encrypted_long = encrypt(long_text, short_key)
    decrypted_long = decrypt(encrypted_long, short_key)
    
    print(f"Encrypted: {encrypted_long}")
    print(f"Decrypted: {decrypted_long}")
    print(f"Success:   {long_text == decrypted_long}")
    
    # Example 5: Case preservation
    print("\n5. Case Preservation")
    print("-" * 40)
    
    mixed_case = "Hello World"
    key = "KEY"
    
    print(f"Text:      {mixed_case}")
    print(f"Key:       {key}")
    
    encrypted_mixed = encrypt(mixed_case, key)
    decrypted_mixed = decrypt(encrypted_mixed, key)
    
    print(f"Encrypted: {encrypted_mixed}")
    print(f"Decrypted: {decrypted_mixed}")
    print(f"Success:   {mixed_case == decrypted_mixed}")
    
    # Example 6: Random key generation
    print("\n6. Random Key Generation")
    print("-" * 40)
    
    plaintext = "RANDOM KEY EXAMPLE"
    
    # Generate random key
    random_key = generate_random_key(10)
    print(f"Random key (length 10): {random_key}")
    
    encrypted_random = encrypt(plaintext, random_key)
    decrypted_random = decrypt(encrypted_random, random_key)
    
    print(f"Plaintext:  {plaintext}")
    print(f"Encrypted:  {encrypted_random}")
    print(f"Decrypted:  {decrypted_random}")
    print(f"Success:    {plaintext == decrypted_random}")
    
    # Generate key for specific text
    auto_key = generate_key_for_text(plaintext)
    print(f"\nAuto-generated key: {auto_key}")
    
    encrypted_auto = encrypt(plaintext, auto_key)
    decrypted_auto = decrypt(encrypted_auto, auto_key)
    
    print(f"Plaintext:  {plaintext}")
    print(f"Encrypted:  {encrypted_auto}")
    print(f"Decrypted:  {decrypted_auto}")
    print(f"Success:    {plaintext == decrypted_auto}")
    
    # Example 7: Encrypt with random key (returns both ciphertext and key)
    print("\n7. Encrypt with Random Key")
    print("-" * 40)
    
    plaintext = "CONFIDENTIAL MESSAGE"
    encrypted_text, generated_key = encrypt_with_random_key(plaintext)
    
    print(f"Plaintext:     {plaintext}")
    print(f"Generated Key: {generated_key}")
    print(f"Encrypted:     {encrypted_text}")
    
    # Decrypt using the generated key
    decrypted_text = decrypt(encrypted_text, generated_key)
    print(f"Decrypted:     {decrypted_text}")
    print(f"Success:       {plaintext == decrypted_text}")
    
    # Example 8: Turkish alphabet support
    print("\n8. Turkish Alphabet Support")
    print("-" * 40)
    
    turkish_alphabet = "ABCÇDEFGĞHIİJKLMNOÖPRSŞTUÜVYZ"
    turkish_text = "MERHABA DÜNYA"
    key = "KEY"
    
    print(f"Turkish Alphabet: {turkish_alphabet}")
    print(f"Turkish Text:     {turkish_text}")
    print(f"Key:              {key}")
    
    encrypted_turkish = encrypt(turkish_text, key, alphabet=turkish_alphabet)
    decrypted_turkish = decrypt(encrypted_turkish, key, alphabet=turkish_alphabet)
    
    print(f"Encrypted:        {encrypted_turkish}")
    print(f"Decrypted:        {decrypted_turkish}")
    print(f"Success:          {turkish_text == decrypted_turkish}")
    
    # Example 9: Error handling
    print("\n9. Error Handling")
    print("-" * 40)
    
    try:
        encrypt("HELLO", "123")  # Invalid key
    except ValueError as e:
        print(f"Invalid key error: {e}")
    
    try:
        encrypt("HELLO", "")  # Empty key
    except ValueError as e:
        print(f"Empty key error: {e}")
    
    try:
        encrypt("HELLO", "AB1C")  # Mixed alphanumeric key
    except ValueError as e:
        print(f"Mixed alphanumeric key error: {e}")
    
    # Example 10: Comparison with other ciphers
    print("\n10. Porta vs Other Ciphers")
    print("-" * 40)
    
    plaintext = "COMPARISON TEST"
    key = "KEY"
    
    print(f"Plaintext: {plaintext}")
    print(f"Key:       {key}")
    
    # Porta encryption
    porta_encrypted = encrypt(plaintext, key)
    print(f"Porta encrypted: {porta_encrypted}")
    
    # Demonstrate self-reciprocal property
    porta_decrypted = decrypt(porta_encrypted, key)
    print(f"Porta decrypted: {porta_decrypted}")
    print(f"Self-reciprocal: {porta_decrypted == plaintext}")
    
    print("\n" + "=" * 60)
    print("PORTA CIPHER EXAMPLE COMPLETED")
    print("=" * 60)


if __name__ == "__main__":
    main()

File: c/examples/porta_example.c
/**
 * @file porta_example.c
 * @brief Example usage of the enhanced Porta cipher with custom pairing support
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "cryptology/classical/substitution/polyalphabetic/porta.h"

#define MAX_BUFFER_SIZE 256
#define MAX_PAIRS 20

int main() {
    printf("======================================================================\n");
    printf("ENHANCED PORTA CIPHER EXAMPLE - CUSTOM PAIRING SUPPORT\n");
    printf("======================================================================\n");
    
    char encrypted[MAX_BUFFER_SIZE];
    char decrypted[MAX_BUFFER_SIZE];
    char key[MAX_BUFFER_SIZE];
    porta_pair_t pairs[MAX_PAIRS];
    size_t pairs_count;
    
    // Example 1: Default pairs
    printf("\n1. Default Alphabet Pairs\n");
    printf("--------------------------------------------------\n");
    
    if (porta_produce_pairs("default", NULL, pairs, &pairs_count, NULL, 0) == 0) {
        printf("Default pairs (first 5): ");
        for (size_t i = 0; i < 5 && i < pairs_count; i++) {
            printf("(%c,%c) ", pairs[i].first, pairs[i].second);
        }
        printf("\nTotal pairs: %zu\n", pairs_count);
        
        const char* plaintext = "HELLO";
        const char* key_str = "KEY";
        
        if (porta_encrypt(plaintext, key_str, NULL, pairs, pairs_count, encrypted, sizeof(encrypted)) == 0) {
            printf("Plaintext: %s\n", plaintext);
            printf("Encrypted: %s\n", encrypted);
            
            if (porta_decrypt(encrypted, key_str, NULL, pairs, pairs_count, decrypted, sizeof(decrypted)) == 0) {
                printf("Decrypted: %s\n", decrypted);
                printf("Success: %s\n", strcmp(plaintext, decrypted) == 0 ? "YES" : "NO");
            }
        }
    }
    
    // Example 2: Turkish alphabet pairs
    printf("\n2. Turkish Alphabet Pairs\n");
    printf("--------------------------------------------------\n");
    
    const char* turkish_alphabet = "ABCÇDEFGĞHIİJKLMNOÖPRSŞTUÜVYZ";
    printf("Turkish alphabet: %s\n", turkish_alphabet);
    
    if (porta_produce_pairs("turkish", turkish_alphabet, pairs, &pairs_count, NULL, 0) == 0) {
        printf("Turkish pairs: ");
        for (size_t i = 0; i < pairs_count; i++) {
            printf("(%c,%c) ", pairs[i].first, pairs[i].second);
        }
        printf("\nTotal Turkish pairs: %zu\n", pairs_count);
        
        const char* turkish_text = "MERHABA";
        const char* key_str = "A";
        
        if (porta_encrypt(turkish_text, key_str, turkish_alphabet, pairs, pairs_count, encrypted, sizeof(encrypted)) == 0) {
            printf("Turkish text: %s\n", turkish_text);
            printf("Encrypted: %s\n", encrypted);
            
            if (porta_decrypt(encrypted, key_str, turkish_alphabet, pairs, pairs_count, decrypted, sizeof(decrypted)) == 0) {
                printf("Decrypted: %s\n", decrypted);
                printf("Success: %s\n", strcmp(turkish_text, decrypted) == 0 ? "YES" : "NO");
            }
        }
    }
    
    // Example 3: Custom user-defined pairs
    printf("\n3. Custom User-Defined Pairs\n");
    printf("--------------------------------------------------\n");
    
    porta_pair_t custom_pairs[] = {
        {'A', 'Z'}, {'B', 'Y'}, {'C', 'X'}, {'D', 'W'}, {'E', 'V'}
    };
    size_t custom_pairs_count = sizeof(custom_pairs) / sizeof(custom_pairs[0]);
    
    printf("Custom pairs: ");
    for (size_t i = 0; i < custom_pairs_count; i++) {
        printf("(%c,%c) ", custom_pairs[i].first, custom_pairs[i].second);
    }
    printf("\n");
    
    if (porta_produce_pairs("custom", NULL, pairs, &pairs_count, custom_pairs, custom_pairs_count) == 0) {
        const char* plaintext = "ABCDE";
        const char* key_str = "ABCDE";
        
        if (porta_encrypt(plaintext, key_str, NULL, pairs, pairs_count, encrypted, sizeof(encrypted)) == 0) {
            printf("Plaintext: %s\n", plaintext);
            printf("Encrypted: %s\n", encrypted);
            
            if (porta_decrypt(encrypted, key_str, NULL, pairs, pairs_count, decrypted, sizeof(decrypted)) == 0) {
                printf("Decrypted: %s\n", decrypted);
                printf("Success: %s\n", strcmp(plaintext, decrypted) == 0 ? "YES" : "NO");
            }
        }
    }
    
    // Example 4: Balanced pairs
    printf("\n4. Balanced Alphabet Pairs\n");
    printf("--------------------------------------------------\n");
    
    const char* alphabet = "ABCDEFGHIJKL";  // 12 letters
    printf("Alphabet: %s\n", alphabet);
    
    if (porta_produce_pairs("balanced", alphabet, pairs, &pairs_count, NULL, 0) == 0) {
        printf("Balanced pairs: ");
        for (size_t i = 0; i < pairs_count; i++) {
            printf("(%c,%c) ", pairs[i].first, pairs[i].second);
        }
        printf("\n");
        
        const char* plaintext = "ABC";
        const char* key_str = "ABC";
        
        if (porta_encrypt(plaintext, key_str, alphabet, pairs, pairs_count, encrypted, sizeof(encrypted)) == 0) {
            printf("Plaintext: %s\n", plaintext);
            printf("Encrypted: %s\n", encrypted);
            
            if (porta_decrypt(encrypted, key_str, alphabet, pairs, pairs_count, decrypted, sizeof(decrypted)) == 0) {
                printf("Decrypted: %s\n", decrypted);
                printf("Success: %s\n", strcmp(plaintext, decrypted) == 0 ? "YES" : "NO");
            }
        }
    }
    
    // Example 5: Self-reciprocal property
    printf("\n5. Self-Reciprocal Property\n");
    printf("--------------------------------------------------\n");
    
    const char* plaintext = "SECRET MESSAGE";
    const char* key_str = "PORTACIPHER";
    
    if (porta_encrypt(plaintext, key_str, NULL, NULL, 0, encrypted, sizeof(encrypted)) == 0) {
        printf("Plaintext: %s\n", plaintext);
        printf("Encrypted: %s\n", encrypted);
        
        // Demonstrate self-reciprocal property
        char encrypted_again[MAX_BUFFER_SIZE];
        if (porta_encrypt(encrypted, key_str, NULL, NULL, 0, encrypted_again, sizeof(encrypted_again)) == 0) {
            printf("Encrypt encrypted text: %s\n", encrypted_again);
            printf("Self-reciprocal: %s\n", strcmp(encrypted_again, plaintext) == 0 ? "YES" : "NO");
        }
    }
    
    // Example 6: Random key generation
    printf("\n6. Random Key Generation\n");
    printf("--------------------------------------------------\n");
    
    const char* random_text = "RANDOM KEY EXAMPLE";
    
    // Generate random key
    if (porta_generate_random_key(10, NULL, key, sizeof(key)) == 0) {
        printf("Random key (length 10): %s\n", key);
        
        if (porta_encrypt(random_text, key, NULL, NULL, 0, encrypted, sizeof(encrypted)) == 0) {
            printf("Plaintext: %s\n", random_text);
            printf("Encrypted: %s\n", encrypted);
            
            if (porta_decrypt(encrypted, key, NULL, NULL, 0, decrypted, sizeof(decrypted)) == 0) {
                printf("Decrypted: %s\n", decrypted);
                printf("Success: %s\n", strcmp(random_text, decrypted) == 0 ? "YES" : "NO");
            }
        }
    }
    
    // Example 7: Encrypt with random key
    printf("\n7. Encrypt with Random Key\n");
    printf("--------------------------------------------------\n");
    
    const char* confidential_text = "CONFIDENTIAL MESSAGE";
    
    if (porta_encrypt_with_random_key(confidential_text, NULL, NULL, 0, 0,
                                     encrypted, sizeof(encrypted),
                                     key, sizeof(key)) == 0) {
        printf("Plaintext: %s\n", confidential_text);
        printf("Generated Key: %s\n", key);
        printf("Encrypted: %s\n", encrypted);
        
        if (porta_decrypt(encrypted, key, NULL, NULL, 0, decrypted, sizeof(decrypted)) == 0) {
            printf("Decrypted: %s\n", decrypted);
            printf("Success: %s\n", strcmp(confidential_text, decrypted) == 0 ? "YES" : "NO");
        }
    }
    
    // Example 8: Error handling
    printf("\n8. Error Handling\n");
    printf("--------------------------------------------------\n");
    
    // Test invalid key
    if (porta_encrypt("HELLO", "123", NULL, NULL, 0, encrypted, sizeof(encrypted)) == -1) {
        printf("Invalid key error: Correctly rejected non-alphabetic key\n");
    }
    
    // Test empty key
    if (porta_encrypt("HELLO", "", NULL, NULL, 0, encrypted, sizeof(encrypted)) == -1) {
        printf("Empty key error: Correctly rejected empty key\n");
    }
    
    // Test invalid pair type
    if (porta_produce_pairs("invalid", NULL, pairs, &pairs_count, NULL, 0) == -1) {
        printf("Invalid pair type error: Correctly rejected invalid pair type\n");
    }
    
    // Test missing custom pairs
    if (porta_produce_pairs("custom", NULL, pairs, &pairs_count, NULL, 0) == -1) {
        printf("Missing custom pairs error: Correctly rejected missing custom pairs\n");
    }
    
    printf("\n======================================================================\n");
    printf("ENHANCED PORTA CIPHER EXAMPLE COMPLETED\n");
    printf("======================================================================\n");
    
    return 0;
}
